test/stack_controller_test.go

```diff
@@ -205,6 +207,386 @@ var _ = Describe("Stack Controller", func() {
 		})
 	})
 
+	Context("configuring a stack using ConfigRefs", func() {
+		var stack *pulumiv1.Stack
+
+		When("using a FileSystemRef", func() {
+			var configDir string
+
+			BeforeEach(func() {
+				By("Creating directory to store configs")
+				configDir, err = os.MkdirTemp("", "secrets")
+				if err != nil {
+					Fail("Failed to create config temp directory")
+				}
+				Expect(os.WriteFile(filepath.Join(configDir, "word.txt"), []byte("just-a-word-in-a-file"), 0600)).To(Succeed())
+			})
+
+			AfterEach(func() {
+				deleteAndWaitForFinalization(stack)
+			})
+
+			It("can deploy a stack reading a config from a file", func() {
+
+				// Use a local backend for this test.
+				// Local backend doesn't allow setting slashes in stack name.
+				const stackName = "dev"
+				fmt.Fprintf(GinkgoWriter, "Stack.Name: %s\n", stackName)
+
+				// Define the stack spec
+				localSpec := shared.StackSpec{
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					Stack:   stackName,
+					GitSource: &shared.GitSource{
+						ProjectRepo: baseDir,
+						RepoDir:     "test/testdata/config-refs",
+						Commit:      commit,
+					},
+					SecretsProvider: "passphrase",
+					EnvRefs:         defaultEnvRefs(),
+					ConfigRefs: map[string]shared.ConfigRef{
+						"word": shared.NewFileSystemConfigResourceRef(filepath.Join(configDir, "word.txt")),
+					},
+					Refresh: true,
+				}
+
+				// Create the stack
+				name := "config-refs-with-file-stack"
+				stack = generateStackV1(name, namespace, localSpec)
+				Expect(k8sClient.Create(ctx, stack)).Should(Succeed())
+
+				// Check that the stack updated successfully
+				fetched := &pulumiv1.Stack{}
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, types.NamespacedName{Name: stack.Name, Namespace: namespace}, fetched)
+					if err != nil {
+						return false
+					}
+					return stackUpdatedToCommit(fetched.Status.LastUpdate, stack.Spec.Commit)
+				}, stackExecTimeout, interval).Should(BeTrue())
+				// Validate outputs.
+				Expect(fetched.Status.Outputs).Should(HaveKeyWithValue("word", v1.JSON{Raw: []byte(`"just-a-word-in-a-file"`)}))
+			})
+		})
+
+		When("using an EnvRef", func() {
+			AfterEach(func() {
+				deleteAndWaitForFinalization(stack)
+			})
+
+			It("can deploy a stack reading a config from an EnvVar", func() {
+
+				// Use a local backend for this test.
+				// Local backend doesn't allow setting slashes in stack name.
+				const stackName = "dev"
+				fmt.Fprintf(GinkgoWriter, "Stack.Name: %s\n", stackName)
+
+				err := os.Setenv("WORD", "just-a-word")
+				if err != nil {
+					Fail("Unable to set WORD environment variable.")
+				}
+
+				// Define the stack spec
+				localSpec := shared.StackSpec{
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					Stack:   stackName,
+					GitSource: &shared.GitSource{
+						ProjectRepo: baseDir,
+						RepoDir:     "test/testdata/config-refs",
```

Comment:
Could we just `test/testdata/structured-config-refs` for all of these? This is testing we can read the configs, seems worth also confirming the structure was handled correctly.

Reply:
Actually this test is trying to read a file, and in a similar way that I implemented with `Literal`, I used the same existing behavior to `FileRef` (the operator already use it to load envvars and secrets from files), which means that just loads a single value. So this test couldn't use `test/testdata/structured-config-refs`

Do you believe that would be worth implementing a `file` behavior specifically to load configs? So we could support structured values as well.